File: kernel/rust/src/fs/devfs.rs
//! Device filesystem - /dev/null, /dev/zero, /dev/random

use alloc::boxed::Box;
use alloc::collections::BTreeMap;
use alloc::string::String;
use alloc::vec::Vec;

use crate::bridge;
use crate::vfs::traits::{
    CallerId, DirEntry, FileHandle, FileSystem, FsError, KPath, Metadata, NodeType, OpenFlags,
    Result, SeekFrom,
};

/// Device types
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DevType {
    Null,   // /dev/null - discards writes, EOF on read
    Zero,   // /dev/zero - infinite zeros
    Random, // /dev/random - random bytes from host
    Cons,   // /dev/cons - the terminal (Plan 9): writes go to stdout, read EOF
}

/// Device filesystem
pub struct DevFs {
    devices: BTreeMap<String, DevType>,
}

impl DevFs {
    pub fn new() -> Self {
        let mut devices = BTreeMap::new();
        devices.insert(String::from("/null"), DevType::Null);
        devices.insert(String::from("/zero"), DevType::Zero);
        devices.insert(String::from("/random"), DevType::Random);
        devices.insert(String::from("/cons"), DevType::Cons);

        Self { devices }
    }

    /// Normalize path
    fn normalize_path(&self, path: &KPath) -> String {
        let s = path.as_str();
        if s.is_empty() {
            String::from("/")
        } else {
            String::from(s)
        }
    }

    /// Get device by path
    fn get_device(&self, path: &str) -> Option<DevType> {
        // Handle both /dev/X and /X
        let clean_path = if path.starts_with("/dev") {
            String::from(&path[4..])
        } else {
            String::from(path)
        };

        self.devices.get(&clean_path).copied()
    }
}

/// File handle for device files
pub struct DevFileHandle {
    dev_type: DevType,
    offset: u64,
}

impl DevFileHandle {
    fn new(dev_type: DevType) -> Self {
        Self {
            dev_type,
            offset: 0,
        }
    }
}

impl FileHandle for DevFileHandle {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        match self.dev_type {
            DevType::Null => {
                // EOF immediately
                Ok(0)
            }
            DevType::Zero => {
                // Fill with zeros
                for byte in buf.iter_mut() {
                    *byte = 0;
                }
                self.offset += buf.len() as u64;
                Ok(buf.len())
            }
            DevType::Random => {
                // Get random bytes from host
                unsafe {
                    bridge::mc_random(buf.as_mut_ptr(), buf.len());
                }
                self.offset += buf.len() as u64;
                Ok(buf.len())
            }
            // The console as a file (Plan 9 `/dev/cons`).
            // Reading returns EOF — terminal input is owned by the line
            // discipline, not a pull-from-fd source.
            DevType::Cons => Ok(0),
        }
    }

    fn write(&mut self, buf: &[u8]) -> Result<usize> {
        match self.dev_type {
            DevType::Null => {
                // Discard all writes
                Ok(buf.len())
            }
            // `/dev/cons` write → the agent's terminal (stdout). Guests write LF;
            // the terminal layer adds the CR (ONLCR), like any other tty output.
            DevType::Cons => {
                crate::io::term_write_stdout(buf);
                Ok(buf.len())
            }
            DevType::Zero | DevType::Random => {
                // Cannot write to these devices
                Err(FsError::PermissionDenied)
            }
        }
    }

    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
        let new_offset = match pos {
            SeekFrom::Start(n) => n as i64,
            SeekFrom::Current(n) => self.offset as i64 + n,
            SeekFrom::End(n) => {
                // Devices have no end
                if n < 0 { 0 } else { n as i64 }
            }
        };

        if new_offset < 0 {
            return Err(FsError::InvalidPath);
        }

        self.offset = new_offset as u64;
        Ok(self.offset)
    }

    fn stat(&self) -> Result<Metadata> {
        let size = match self.dev_type {
            DevType::Null | DevType::Cons => 0,
            DevType::Zero | DevType::Random => 0, // Infinite/variable size devices
        };

        Ok(Metadata::file(size))
    }
}

impl FileSystem for DevFs {
    fn open(
        &mut self,
        path: &KPath,
        flags: OpenFlags,
        _caller: CallerId,
    ) -> Result<Box<dyn FileHandle>> {
        let path_str = self.normalize_path(path);

        // Get device type
        let dev_type = self.get_device(&path_str).ok_or(FsError::NotFound)?;

        // Check write permissions
        if flags.write {
            match dev_type {
                DevType::Zero | DevType::Random => {
                    return Err(FsError::PermissionDenied);
                }
                // Null discards writes; cons forwards them to the terminal.
                DevType::Null | DevType::Cons => {}
            }
        }

        let handle = DevFileHandle::new(dev_type);
        Ok(Box::new(handle))
    }

    fn stat(&self, path: &KPath) -> Result<Metadata> {
        let path_str = self.normalize_path(path);
        if path_str == "/" || path_str == "/dev" {
            return Ok(Metadata::dir());
        }

        let dev_type = self.get_device(&path_str).ok_or(FsError::NotFound)?;

        let size = match dev_type {
            DevType::Null | DevType::Cons => 0,
            DevType::Zero | DevType::Random => 0,
        };

        Ok(Metadata::file(size))
    }

    fn readdir(&self, _path: &KPath, _caller: CallerId) -> Result<Vec<DirEntry>> {
        // List all devices
        let mut result = Vec::new();

        for (name, _dev_type) in &self.devices {
            let display_name = if name.starts_with('/') {
                String::from(&name[1..]) // Remove leading /
            } else {
                name.clone()
            };

            result.push(DirEntry {
                name: display_name,
                node_type: NodeType::File,
            });
        }

        Ok(result)
    }

    fn mkdir(&mut self, _path: &KPath, _caller: CallerId) -> Result<()> {
        Err(FsError::PermissionDenied) // Cannot create directories in /dev
    }

    fn unlink(&mut self, _path: &KPath, _caller: CallerId) -> Result<()> {
        Err(FsError::PermissionDenied) // Cannot delete devices
    }

    fn rename(&mut self, _from: &KPath, _to: &KPath, _caller: CallerId) -> Result<()> {
        Err(FsError::PermissionDenied) // Cannot rename devices
    }
}

File: memcontainers/kernel/zig/src/fs/devfs.zig
//! src/fs/devfs.zig — device files: /dev/null, /dev/zero, /dev/random, /dev/cons (§2.5).
//!
//! Owns: the device nodes and their read/write behavior; entropy and console output flow
//! through the bridge (A5). Oracle: kernel/rust/src/fs/devfs.rs.
//! Not here: the line editor (scheduler.zig); pipes (ipc/pipe.zig).

const std = @import("std");
const vfs = @import("../vfs.zig");
const bridge = @import("../bridge.zig");
const FsError = vfs.FsError;
const NodeType = vfs.NodeType;
const Metadata = vfs.Metadata;
const OpenFlags = vfs.OpenFlags;
const SeekFrom = vfs.SeekFrom;
const FileHandle = vfs.FileHandle;
const FileSystem = vfs.FileSystem;
const DirEntry = vfs.DirEntry;

const DevType = enum { null_, zero, random, cons };

const DEVICES = [_]struct { name: []const u8, dev: DevType }{
    .{ .name = "null", .dev = .null_ },
    .{ .name = "zero", .dev = .zero },
    .{ .name = "random", .dev = .random },
    .{ .name = "cons", .dev = .cons },
};

pub const DevFs = struct {
    gpa: std.mem.Allocator,

    pub fn create(gpa: std.mem.Allocator) *DevFs {
        const self = gpa.create(DevFs) catch @panic("OOM");
        self.* = .{ .gpa = gpa };
        return self;
    }
    pub fn fileSystem(self: *DevFs) FileSystem {
        return .{ .ptr = self, .vtable = &fs_vtable };
    }

    /// Handle both /dev/X and /X (the namespace hands us the mount-relative path).
    fn getDevice(path: []const u8) ?DevType {
        const clean = if (std.mem.startsWith(u8, path, "/dev")) path[4..] else path;
        const name = if (std.mem.startsWith(u8, clean, "/")) clean[1..] else clean;
        for (DEVICES) |d| {
            if (std.mem.eql(u8, d.name, name)) return d.dev;
        }
        return null;
    }

    fn open(self: *DevFs, path: []const u8, flags: OpenFlags) FsError!FileHandle {
        const dev = getDevice(path) orelse return FsError.NotFound;
        if (flags.write) switch (dev) {
            .zero, .random => return FsError.PermissionDenied,
            .null_, .cons => {},
        };
        const h = self.gpa.create(DevFileHandle) catch @panic("OOM");
        h.* = .{ .fs = self, .dev = dev, .offset = 0 };
        return .{ .ptr = h, .vtable = &DevFileHandle.handle_vtable };
    }

    fn stat(_: *DevFs, path: []const u8) FsError!Metadata {
        if (std.mem.eql(u8, path, "/") or std.mem.eql(u8, path, "/dev")) return Metadata.dir();
        _ = getDevice(path) orelse return FsError.NotFound;
        return Metadata.file(0);
    }

    fn readdir(_: *DevFs, _: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(DirEntry)) FsError!void {
        for (DEVICES) |d| {
            out.append(arena, .{ .name = arena.dupe(u8, d.name) catch @panic("OOM"), .node_type = .file }) catch @panic("OOM");
        }
    }

    const fs_vtable = FileSystem.VTable{
        .open = fsOpen,
        .stat = fsStat,
        .readdir = fsReaddir,
        .mkdir = fsReadOnly1,
        .unlink = fsReadOnly1,
        .rename = vfs.fsRenameUnsupported,
        .symlink = vfs.fsSymlinkUnsupported,
        .link = vfs.fsLinkUnsupported,
        .readlink = vfs.fsReadlinkUnsupported,
        .setMode = vfs.fsSetModeUnsupported,
        .setTimes = vfs.fsSetTimesUnsupported,
    };
    fn fsOpen(p: *anyopaque, _: vfs.CallerId, path: []const u8, flags: OpenFlags) FsError!FileHandle {
        return self_(p).open(path, flags);
    }
    fn fsStat(p: *anyopaque, path: []const u8) FsError!Metadata {
        return self_(p).stat(path);
    }
    fn fsReaddir(p: *anyopaque, _: vfs.CallerId, path: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(DirEntry)) FsError!void {
        return self_(p).readdir(path, arena, out);
    }
    fn fsReadOnly1(_: *anyopaque, _: vfs.CallerId, _: []const u8) FsError!void {
        return FsError.PermissionDenied;
    }
    fn self_(p: *anyopaque) *DevFs {
        return @ptrCast(@alignCast(p));
    }
};

const DevFileHandle = struct {
    fs: *DevFs,
    dev: DevType,
    offset: u64,

    fn read(self: *DevFileHandle, buf: []u8) FsError!usize {
        switch (self.dev) {
            .null_, .cons => return 0,
            .zero => {
                @memset(buf, 0);
                self.offset += buf.len;
                return buf.len;
            },
            .random => {
                bridge.mc_random(buf.ptr, buf.len);
                self.offset += buf.len;
                return buf.len;
            },
        }
    }
    fn write(self: *DevFileHandle, buf: []const u8) FsError!usize {
        switch (self.dev) {
            .null_ => return buf.len,
            .cons => {
                bridge.mc_stdout_write(buf.ptr, buf.len);
                return buf.len;
            },
            .zero, .random => return FsError.PermissionDenied,
        }
    }
    fn seek(self: *DevFileHandle, pos: SeekFrom) FsError!u64 {
        const new_off: i64 = switch (pos) {
            .start => |n| @intCast(n),
            .current => |n| @as(i64, @intCast(self.offset)) + n,
            .end => |n| if (n < 0) 0 else n,
        };
        if (new_off < 0) return FsError.InvalidPath;
        self.offset = @intCast(new_off);
        return self.offset;
    }
    fn stat(_: *DevFileHandle) FsError!Metadata {
        return Metadata.file(0);
    }

    const handle_vtable = FileHandle.VTable{
        .read = hRead,
        .write = hWrite,
        .seek = hSeek,
        .stat = hStat,
        .truncate = vfs.handleTruncateUnsupported,
        .close = hClose,
    };
    fn hRead(p: *anyopaque, buf: []u8) FsError!usize {
        return h_(p).read(buf);
    }
    fn hWrite(p: *anyopaque, buf: []const u8) FsError!usize {
        return h_(p).write(buf);
    }
    fn hSeek(p: *anyopaque, pos: SeekFrom) FsError!u64 {
        return h_(p).seek(pos);
    }
    fn hStat(p: *anyopaque) FsError!Metadata {
        return h_(p).stat();
    }
    fn hClose(p: *anyopaque) void {
        const self = h_(p);
        self.fs.gpa.destroy(self);
    }
    fn h_(p: *anyopaque) *DevFileHandle {
        return @ptrCast(@alignCast(p));
    }
};
